cli/tools/fmt/incremental_cache.rs

```diff
@@ -0,0 +1,294 @@
+use std::collections::HashMap;
+use std::path::Path;
+use std::path::PathBuf;
+
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+use deno_core::serde_json;
+use rusqlite::params;
+use rusqlite::Connection;
+use tokio::task::JoinHandle;
+
+use crate::config_file::FmtOptionsConfig;
+
+enum ReceiverMessage {
+  Update(PathBuf, u64),
+  Exit,
+}
+
+/// Cache used to not bother formatting a file again when we
+/// know it is already formatted.
+pub struct IncrementalCache {
+  previous_hashes: HashMap<PathBuf, u64>,
+  sender: tokio::sync::mpsc::UnboundedSender<ReceiverMessage>,
+  handle: Mutex<Option<JoinHandle<()>>>,
+}
+
+impl IncrementalCache {
+  pub fn new(
+    db_file_path: &Path,
+    fmt_config: &FmtOptionsConfig,
+    initial_file_paths: &[PathBuf],
+  ) -> Result<Self, AnyError> {
+    let state_hash =
+      fast_insecure_hash(serde_json::to_string(fmt_config).unwrap().as_bytes());
+    let sql_cache = SqlIncrementalCache::new(db_file_path, state_hash)?;
+    Ok(Self::from_sql_incremental_cache(
+      sql_cache,
+      initial_file_paths,
+    ))
+  }
+
+  fn from_sql_incremental_cache(
+    cache: SqlIncrementalCache,
+    initial_file_paths: &[PathBuf],
+  ) -> Self {
+    let mut previous_hashes = HashMap::new();
+    for path in initial_file_paths {
+      if let Some(hash) = cache.get_source_hash(path) {
+        previous_hashes.insert(path.to_path_buf(), hash);
+      }
+    }
+
+    let (sender, mut receiver) =
+      tokio::sync::mpsc::unbounded_channel::<ReceiverMessage>();
+
+    // sqlite isn't `Sync`, so we do all the updating on a dedicated thread
+    let handle = tokio::task::spawn(async move {
```

Comment:
`tokio::task::spawn` doesn't create a dedicated thread, it creates a new task that will be run on the same runtime instance (ie. main thread)

Reply:
I think that's ok because all the work is being done on blocking threads? I should update the comment to remove the word thread.

Reply:
Yes, I think that is fine